pkg/util/git.go

```diff
@@ -0,0 +1,276 @@
+	if len(split) == 2 {
+		splitFile = strings.SplitN(split[1], "/", 3)
+	}
+
+	if len(splitOrg) < 2 {
+		err = fmt.Errorf("url path should contain <user>/<repo>, received: %s", url.Path[1:])
+	} else {
+		g.Owner = splitOrg[0]
+		g.Repo = splitOrg[1]
+	}
+
+	if len(splitFile) == 3 {
+		if splitFile[0] == "blob" || splitFile[0] == "tree" || splitFile[0] == "raw" {
+			g.Branch = splitFile[1]
+			g.Path = splitFile[2]
+			ext := filepath.Ext(g.Path)
+			if ext != "" {
+				g.IsFile = true
+			}
+		} else {
+			err = fmt.Errorf("url path should contain 'blob' or 'tree' or 'raw', received: %s", url.Path[1:])
+		}
+	}
+
+	if !isGitUrlPublic(g, c) {
+		g.token = os.Getenv(GitLabToken)
+	}
+
+	return g, err
+}
+
+func parseBitbucketUrl(g GitUrl, url *url.URL, c http.Client) (GitUrl, error) {
+	var splitUrl []string
+	var err error
+
+	g.Protocol = url.Scheme
+	g.Host = url.Host
+	g.IsFile = false
+
+	splitUrl = strings.SplitN(url.Path[1:], "/", 5)
+	if len(splitUrl) < 2 {
+		err = fmt.Errorf("url path should contain <user>/<repo>, received: %s", url.Path[1:])
+	} else if len(splitUrl) == 2 {
+		g.Owner = splitUrl[0]
+		g.Repo = splitUrl[1]
+	} else {
+		g.Owner = splitUrl[0]
+		g.Repo = splitUrl[1]
+		if len(splitUrl) == 5 {
+			if splitUrl[2] == "raw" || splitUrl[2] == "src" {
+				g.Branch = splitUrl[3]
+				g.Path = splitUrl[4]
+				ext := filepath.Ext(g.Path)
+				if ext != "" {
+					g.IsFile = true
+				}
+			} else {
+				err = fmt.Errorf("url path should contain 'raw' or 'src', received: %s", url.Path[1:])
+			}
+		} else {
+			err = fmt.Errorf("url path should contain path to directory or file, received: %s", url.Path[1:])
+		}
+	}
+
+	if !isGitUrlPublic(g, c) {
+		g.token = os.Getenv(BitbucketToken)
+	}
+
+	return g, err
+}
+
+func isGitUrlPublic(g GitUrl, c http.Client) bool {
+	host := g.Host
+	if host == RawGitHubHost {
+		host = GitHubHost
+	}
+
+	repo := fmt.Sprintf("%s://%s/%s/%s", g.Protocol, host, g.Owner, g.Repo)
+
+	if res, err := c.Get(repo); err != nil {
+		return false
+	} else if res.StatusCode == http.StatusOK {
+		return true
+	}
+	return false
+}
+
+// CloneGitRepo clones a GitHub Repo to a destination directory
+func CloneGitRepo(g GitUrl, destDir string) error {
+	var cloneOptions *gitpkg.CloneOptions
+
+	host := g.Host
+	if host == RawGitHubHost {
+		host = GitHubHost
+	}
+
+	repoUrl := fmt.Sprintf("%s://%s/%s/%s.git", g.Protocol, host, g.Owner, g.Repo)
+	branch := fmt.Sprintf("refs/heads/%s", g.Branch)
+
+	cloneOptions = &gitpkg.CloneOptions{
+		URL:           repoUrl,
+		ReferenceName: plumbing.ReferenceName(branch),
+		SingleBranch:  true,
+		Depth:         1,
+	}
+
+	if g.token != "" {
+		cloneOptions.Auth = &githttp.BasicAuth{
+			// go-git auth allows username to be anything except
+			// an empty string for GitHub and GitLab, however requires
+			// for Bitbucket to be "x-token-auth"
+			Username: "x-token-auth",
+			Password: g.token,
+		}
+	}
+
+	_, err := gitpkg.PlainClone(destDir, false, cloneOptions)
```

Comment:
We had problems using the go-git cloning function in HAS for large repos.  See https://github.com/redhat-appstudio/application-service/pull/204.  It'll be good if we can adopt the same workaround to use the git cli instead

Reply:
Copied the workaround and it looks good